README.md
## Tori.fi Notifier
This a very simple application that enables one to spot new product updates on [Tori.fi](https://www.tori.fi/).  

The current version (v0.2) of this application is command line only, and is only suitable for a 
very specific use-case (firewood purchases at Päijät Häme). 
This is because the current version is merely a prototype.

I might or might not keep expanding this prototype further. So far this has just been a quick side project for me.

### Background - How this app came to be?
Because of the current (2022) energy "crisis", the demand for firewood has increased, making [Tori.fi ](https://www.tori.fi/)
sales (and giveaway) items more sought after. In order to get a good deal, it is necessary to be early on 
the markets, but who wants to keep manually checking the site?

I kind of accidentally challenged myself to create this application, as I was talking with my wife about this 
situation. So, here we are.

In its current form this application merely meets my needs on this area: It looks for new firewood sales and 
opens the browser for me when it has found one. There are no custom parameters, no fancy UI or anything else, just 
the MVP implementation.

### Usage
Build the artifact using [IntelliJ IDEA](https://www.jetbrains.com/idea/). 
Run the Tori-Notifier.jar using command line with: `java -jar Tori-Notifier.jar` 
([JRE 8](https://www.openlogic.com/openjdk-downloads?field_java_parent_version_target_id=416&field_operating_system_target_id=All&field_architecture_target_id=All&field_java_package_target_id=401) 
or later is required)

When you want to close the app, just press enter (i.e. provide any text input).

### Future Improvements
If I end up working on this application, I'll attempt to work on the following features:
1. Customizable search
2. Concurrent searches
3. Graphical user interface

Changes.md
## Tori Notifier - List of Changes

### v0.2 - Release pending
Initial prototype version

src/vf/tori/model/SalesItem.scala
package vf.tori.model

import utopia.flow.datastructure.immutable.{Model, ModelDeclaration}
import utopia.flow.generic.{FromModelFactoryWithSchema, ModelConvertible, StringType}
import utopia.flow.generic.ValueConversions._
import utopia.flow.generic.ValueUnwraps._

object SalesItem extends FromModelFactoryWithSchema[SalesItem]
{
	override val schema = ModelDeclaration("name" -> StringType)
	
	override protected def fromValidatedModel(model: Model) = apply(model("name"), model("price"))
}

/**
 * Represents an item on sale (or being given away)
 * @author Mikko Hilpinen
 * @since 11.9.2022, v0.2
 * @param name Name of this item
 * @param price Price of this item in euros, if known
 */
case class SalesItem(name: String, price: Option[Double] = None) extends ModelConvertible
{
	override def toModel = Model.from("name" -> name, "price" -> price)
	
	override def toString = price match {
		case Some(price) => s"$name: $price €"
		case None => name
	}
}


src/vf/tori/controller/SalesItems.scala
package vf.tori.controller

import utopia.flow.async.LoopingProcess
import utopia.flow.container.ObjectsFileContainer
import utopia.flow.container.SaveTiming.OnJvmClose
import utopia.flow.time.Now
import utopia.flow.time.TimeExtensions._
import utopia.flow.util.CollectionExtensions._
import utopia.flow.util.FileExtensions._
import vf.tori.model.SalesItem
import vf.tori.util.Common._

import java.awt.Desktop
import java.net.URI
import scala.util.{Failure, Success, Try}

/**
 * Caches and manages sales item data locally
 * @author Mikko Hilpinen
 * @since 11.9.2022, v0.2
 */
object SalesItems
{
	// ATTRIBUTES   -------------------------
	
	private val container = new ObjectsFileContainer("data/sales-items.json", SalesItem, OnJvmClose)
	private lazy val updateLoop = LoopingProcess.static(3.minutes) { _ =>
		ReadSalesItems("polttopuu").onComplete { _.flatten match {
			case Success(items) => container.current = items
			case Failure(error) => log(error, "Sales items reading failed")
		} }
	}
	
	
	// INITIAL CODE -------------------------
	
	// Opens the browser every time new items are added
	pointer.addListener { change =>
		if (change.newValue.exists { item => !change.oldValue.contains(item) }) {
			val result = {
				if (Desktop.isDesktopSupported) {
					val desktop = Desktop.getDesktop
					if (desktop.isSupported(Desktop.Action.BROWSE))
						Try { desktop.browse(new URI(
							"https://www.tori.fi/uusimaa?q=polttopuu&cg=0&w=112&st=s&st=g&ca=18&l=0&md=th")) }
					else
						Failure(new UnsupportedOperationException("Internet browsing is not supported in desktop"))
				}
				else
					Failure(new UnsupportedOperationException("Desktop operations are not supported"))
			}
			result.failure.foreach { log(_, "Couldn't open the browser") }
		}
		
		// Also prints updates to console
		println("\n---")
		println(s"${ Now.toLocalDateTime }: Sales items updated")
		change.newValue.foreach { item => println(s"\t- $item") }
		println("---")
	}
	
	
	// COMPUTED -----------------------------
	
	/**
	 * @return The current sales items data
	 */
	def current = container.current
	/**
	 * @return Pointer to the current sales items data
	 */
	def pointer = container.pointer.valueView
	
	
	// OTHER    -----------------------------
	
	/**
	 * Starts automated background updates
	 */
	def startUpdates() = updateLoop.runAsync()
}


src/vf/tori/controller/ReadSalesItems.scala
package vf.tori.controller

import utopia.access.http.StatusGroup
import utopia.bunnymunch.jawn.JsonBunny
import utopia.disciple.apache.Gateway
import utopia.disciple.http.request.Request
import utopia.disciple.http.response.ResponseParser
import utopia.disciple.model.error.RequestFailedException
import utopia.flow.async.AsyncExtensions._
import utopia.flow.generic.ValueConversions._
import utopia.flow.parse.Regex
import utopia.flow.util.IterateLines
import utopia.flow.util.logging.Logger
import utopia.flow.util.CollectionExtensions._
import utopia.flow.util.StringExtensions._
import vf.tori.model.SalesItem

import java.nio.charset.StandardCharsets
import scala.collection.immutable.VectorBuilder
import scala.concurrent.ExecutionContext
import scala.io.Codec
import scala.util.Failure

/**
 * Reads items being given away, or on sale from Tori.fi
 * @author Mikko Hilpinen
 * @since 11.9.2022, v0.2
 */
object ReadSalesItems
{
	private val divStartRegex = Regex("\\<div")
	private val divEndRegex = Regex("\\<\\/div\\>")
	
	private val externalGateway = Gateway(Vector(JsonBunny),
		allowBodyParameters = false, allowJsonInUriParameters = false)
	
	/**
	 * Performs a sales item search
	 * @param search The search string (e.g. "polttopuu")
	 * @param exc Implicit execution context
	 * @param log Implicit logger
	 * @return Items on sale (async). May contain a failure.
	 */
	def apply(search: String)(implicit exc: ExecutionContext, log: Logger) = {
		externalGateway.responseFor(
			Request(s"https://www.tori.fi/uusimaa?q=$search&cg=0&w=112&st=s&st=g&ca=18&l=0&md=th"))(
			ResponseParser.defaultOnEmpty[Vector[SalesItem]](Vector()) { (stream, headers, status) =>
				if (status.group == StatusGroup.Success) {
					implicit val codec: Codec = Codec(headers.charset.getOrElse(StandardCharsets.ISO_8859_1))
					IterateLines.fromStream(stream) { linesIter =>
						val iter = linesIter.pollable
						// Finds a line with either "Ei tuloksia" (terminator)
						// or "class="add-details-left"" (indicating a sales item)
						Iterator.continually {
							iter.nextWhere { line => line.contains("Ei tuloksia") || line.contains("ad-details-left") }
								.flatMap { firstLine =>
									// Case: No results found => terminates
									if (firstLine.contains("Ei tuloksia"))
										None
									// Case: Sales item found => processes it
									else {
										// Parses the whole div
										val divLines = divFrom(firstLine, iter)
										val title = divLines.find { _.contains("li-title") } match {
											case Some(line) => line.afterFirst(">").untilFirst("<")
											case None => ""
										}
										val price = divLines.find { _.contains("list_price ineuros") } match {
											case Some(line) => line.afterFirst(">").untilFirst("<")
											case None => ""
										}
										Some(SalesItem(title, price.untilFirst("&").trim.double))
									}
								}
						}.takeWhile { _.isDefined }.flatten.toVector
					}
				}
				else
					Failure(new RequestFailedException(s"Service responded with $status"))
			})
			.tryMapIfSuccess { response =>
				if (response.isSuccess)
					response.body
				else
					Failure(new RequestFailedException(s"Service responded with status ${ response.status }"))
			}
	}
	
	private def divFrom(firstLine: String, linesIter: Iterator[String]) = {
		// Collects the right amount of lines
		var currentDepth = 1
		val linesBuilder = new VectorBuilder[String]()
		linesBuilder += firstLine
		while (currentDepth > 0 && linesIter.hasNext) {
			val line = linesIter.next()
			currentDepth += divStartRegex.startIndexIteratorIn(line).size - divEndRegex.startIndexIteratorIn(line).size
			/*
			if (line.contains("<div"))
				currentDepth += 1
			else if (line.contains("</div>"))
				currentDepth -= 1
			 */
			linesBuilder += line
		}
		
		linesBuilder.result()
	}
}


src/vf/tori/app/StatusReadTestApp.scala
package vf.tori.app

import utopia.access.http.{Status, StatusGroup}
import utopia.disciple.apache.Gateway
import utopia.disciple.http.request.Request
import utopia.disciple.http.response.ResponseParser
import utopia.disciple.model.error.RequestFailedException
import utopia.flow.async.AsyncExtensions._
import utopia.flow.async.ThreadPool
import utopia.flow.generic.DataType
import utopia.flow.parse.Regex
import utopia.flow.util.IterateLines
import utopia.flow.util.logging.{Logger, SysErrLogger}
import utopia.flow.util.CollectionExtensions._
import utopia.flow.util.StringExtensions._

import java.nio.charset.StandardCharsets
import scala.collection.immutable.VectorBuilder
import scala.concurrent.ExecutionContext
import scala.io.Codec
import scala.util.{Failure, Success}

/**
 * A test application that reads sales items
 * @author Mikko Hilpinen
 * @since 11.9.2022, v0.1
 */
object StatusReadTestApp extends App
{
	DataType.setup()
	Status.setup()
	
	implicit val log: Logger = SysErrLogger
	implicit val exc: ExecutionContext = new ThreadPool("Tori-Notifier").executionContext
	
	val divStartRegex = Regex("\\<div")
	val divEndRegex = Regex("\\<\\/div\\>")
	
	val externalGateway = Gateway(allowBodyParameters = false, allowJsonInUriParameters = false)
	
	println("Requesting data...")
	externalGateway.responseFor(
		Request("https://www.tori.fi/uusimaa?q=polttopuu&cg=0&w=112&st=s&st=g&ca=18&l=0&md=th"))(
		ResponseParser.defaultOnEmpty[Vector[(String, String)]](Vector()) { (stream, headers, status) =>
			if (status.group == StatusGroup.Success) {
				implicit val codec: Codec = Codec(headers.charset.getOrElse(StandardCharsets.ISO_8859_1))
				println("Successful response received. Reading...")
				IterateLines.fromStream(stream) { linesIter =>
					val iter = linesIter.pollable
					// Finds a line with either "Ei tuloksia" (terminator)
					// or "class="add-details-left"" (indicating a sales item)
					Iterator.continually {
						iter.nextWhere { line => line.contains("Ei tuloksia") || line.contains("ad-details-left") }
							.flatMap { firstLine =>
								println(s"Key line found: $firstLine")
								// Case: No results found => terminates
								if (firstLine.contains("Ei tuloksia")) {
									println("No results found")
									None
								}
								// Case: Sales item found => processes it
								else {
									// Parses the whole div
									val divLines = divFrom(firstLine, iter)
									val title = divLines.find { _.contains("li-title") } match {
										case Some(line) => line.afterFirst(">").untilFirst("<")
										case None => ""
									}
									val price = divLines.find { _.contains("list_price ineuros") } match {
										case Some(line) => line.afterFirst(">").untilFirst("<")
										case None => ""
									}
									println(s"Found item:")
									divLines.foreach(println)
									Some(title -> price)
								}
							}
					}.takeWhile { _.isDefined }.flatten.toVector
				}
			}
			else
				Failure(new RequestFailedException(s"Service responded with $status"))
		})
		.waitForResult() match {
			case Success(response) =>
				response.body match {
					case Success(items) =>
						println(s"\n${ items.size } items available:")
						items.foreach { case (title, price) => println(s"$title: $price") }
					case Failure(error) => log(error, response.status.toString)
				}
			case Failure(error) => log(error)
		}
	
	private def divFrom(firstLine: String, linesIter: Iterator[String]) = {
		// Collects the right amount of lines
		var currentDepth = 1
		val linesBuilder = new VectorBuilder[String]()
		linesBuilder += firstLine
		while (currentDepth > 0 && linesIter.hasNext) {
			val line = linesIter.next()
			currentDepth += divStartRegex.startIndexIteratorIn(line).size - divEndRegex.startIndexIteratorIn(line).size
			/*
			if (line.contains("<div"))
				currentDepth += 1
			else if (line.contains("</div>"))
				currentDepth -= 1
			 */
			linesBuilder += line
		}
		
		linesBuilder.result()
	}
}
